### PR TITLE
Refactor/arch components

### PR DESCRIPTION
**Pull request by LiamClark**
Big ass pr that:

1. throws away dead views from the trust chain devs.
2. adds a view pager with tabs to switch between screens instead of the devices button.
3. makes the network operations async, although slightly hacky. this is needed because
before the application would throw network on main thread exceptions.
4. establishes android architecture components for communication between fragments.
5. Actually, allow for peer selection on the network and highlight the selected peer.

### CODE REVIEW

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import java.security.KeyPair
+import java.util.*
+
+class BlockSignerValidator(val storage: TrustChainStorage, val me: ChainService.Peer,
+                           val keyPair: KeyPair) {
+    val myPublicKey = keyPair.public.encoded
+
+    fun signBlock(peer: ChainService.Peer, linkedBlock: MessageProto.TrustChainBlock): MessageProto.TrustChainBlock? {
+        // do nothing if linked block is not addressed to me
+        val myPublicKey = me.publicKey.toByteArray()
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt]: do `keyPair.public.encoded` and `my.publicKey` differ? if so, shadowing this variable might not be the best idea.

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        // do nothing if block is not a request
+        if (linkedBlock.linkSequenceNumber != TrustChainBlock.UNKNOWN_SEQ) {
+            Log.e(TAG, "signBlock: Block is not a request.")
+            return null
+        }
+        var block = TrustChainBlock.createBlock(null, storage,
+                myPublicKey,
+                linkedBlock, peer.publicKey.toByteArray())
+
+        block = TrustChainBlock.sign(block, keyPair.private)
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt:32]: Does the private key only exist in the keyPair object?

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        val validation: ValidationResult?
+        try {
+            validation = TrustChainBlock.validate(block, storage)
+        } catch (e: Exception) {
+            e.printStackTrace()
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt]: Logger?

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                    + validation.getErrors().toString())
+        } else {
+            storage.insertInDB(block)
+            return block
+        }
+        return null
+    }
+
+
+    fun createNewBlock(transaction: ByteArray, peerKey: ByteArray): MessageProto.TrustChainBlock? {
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt]: This is duplicating a lot from the method above, can we re-use this better?

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        Log.i(TAG, "Received half block from peer with IP: " + peer.hostname + ":" + peer.port +
+                " and public key: " + Peer.bytesToHex(peer.publicKey.toByteArray()))
+
+        if (block.linkPublicKey != null && block.publicKey != null) {
+
+            val validation: ValidationResult
+            try {
+                validation = TrustChainBlock.validate(block, storage)
+            } catch (e: Exception) {
+                e.printStackTrace()
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/BlockSignerValidator.kt:84]: => Logger

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainClientRegistry.kt [hunk 6]
```diff
@@ -1,48 +1,18 @@
...
-    }
-
-    fun channelForPeer2(peer: ChainService.Peer): ChainGrpc.ChainStub {
+    private fun channelForPeer(peer: ChainService.Peer): ChainGrpc.ChainFutureStub {
         val channel = ManagedChannelBuilder.forAddress(peer.hostname, peer.port).usePlaintext(true).build();
-        return ChainGrpc.newStub(channel)
+        return ChainGrpc.newFutureStub(channel)
     }
 
-    fun foo(peer: ChainService.Peer, request: ChainService.PeerCrawlRequest): Flowable<ChainService.PeerTrustChainBlock> {
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainClientRegistry.kt:29]: :hankey:

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt [hunk 7]
```diff
@@ -82,14 +93,14 @@ class ChainServiceServer(val storage: TrustChainStorage, val me: ChainService.Pe
...
+    fun verifyExistingBlock(peer: ChainService.Peer, seqNum: Int): Single<Boolean> = storage.getBlock(peer.publicKey.toByteArray(), seqNum)?.let { verifyNewBlock(peer, it) } ?: Single.just(false)
 
-    fun verifyNewBlock(peer: ChainService.Peer, block: MessageProto.TrustChainBlock): Boolean {
+    fun verifyNewBlock(peer: ChainService.Peer, block: MessageProto.TrustChainBlock): Single<Boolean> {
         val publicKey = peer.publicKey.toByteArray()
         val publicResult: SetupPublicResult = block.asSetupPublic() ?: throw IllegalArgumentException("Block did not contain attestation")
         val rangeProofVerifier = RangeProofVerifier(publicResult.N, publicResult.a, publicResult.b)
 
-        val resultOfAllProofs = (0..10).map {
+        val resultOfAllProofs: List<Single<Boolean>> = (0..10).map {
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:103]: `(0..10)`: Was already there, but this gives a 0.999 certainty, don't we want that to run more often? I guess it's a perf issue atm, we should think about how secure we want it to be

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt [hunk 8]
```diff
@@ -204,126 +173,57 @@ class ChainServiceServer(val storage: TrustChainStorage, val me: ChainService.Pe
...
-            Log.e(TAG, "Signed block did not validate. Result: " + validation.toString() + ". Errors: "
-                    + validation.getErrors().toString())
-        } else {
-            storage.insertInDB(block)
-            return block
+        val peerChannel = registry.findStub(peerMessage)
+        val keySingle: Single<ChainService.Key> = peerChannel.sendLatestBlocks(crawlResponse).guavaAsSingle(Schedulers.computation())
+
+        return keySingle.flatMap { theirPublicKey ->
+            val newBlock = signerValidator.createNewBlock(payload, theirPublicKey.publicKey.toByteArray())
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:209]: `theirPublicKey.publicKey`: is theirPublicKey a Keypair by chance?

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt [hunk 9]
```diff
@@ -204,126 +173,57 @@ class ChainServiceServer(val storage: TrustChainStorage, val me: ChainService.Pe
...
+
+            if (newBlock != null) {
+                val trustChainBlock = ChainService.PeerTrustChainBlock.newBuilder().setBlock(newBlock).setPeer(me).build()
+                val recievedCompleteBlock: Single<ChainService.PeerTrustChainBlock> = peerChannel.recieveHalfBlock(trustChainBlock).guavaAsSingle(Schedulers.computation())
+                recievedCompleteBlock.map {
+                    signerValidator.saveCompleteBlock(it)
+                    Log.i(TAG, "saved a block")
+                    it
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:217]: couldn't remember the `for` loop syntax again heh?

File: app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainActivity.kt [hunk 10]
```diff
@@ -1,168 +1,37 @@
...
-
-        val (server, grpc) = ChainServiceServer.createServer(kp, 8080, localIPAddress!!, dbHelper, this::attestationPrompt, zkp.second)
-
-//        val (server, grpc) = ChainServiceServer.createServer(kp, 8080, localIPAddress!!, TrustChainMemoryStorage(kp), this::attestationPrompt, zkp.second)
-        this.server = server
-
-    }
 
-    private fun attestationPrompt(attestation: ChainService.PublicSetupResult): Single<Boolean> {
-        val fuckyiou = Single.create<Boolean> { source ->
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainActivity.kt:115]: :joy:

File: app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+                }
+            } catch (e: Exception) {
+                e.printStackTrace()
+            }
+
+            return null
+        }
+
+    private fun attestationPrompt(attestation: ChainService.PublicSetupResult): Single<Boolean> {
+        val fuckyiou = Single.create<Boolean> { source ->
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainViewModel.kt]: oh it still exists

File: app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+            val message = "attest for some cool dude that his _ is between ${attestation.a} - ${attestation.b}"
```
Comment (GijsWeterings) [inline app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainViewModel.kt]: as does this :P

File: app/src/test/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServerTest.kt [hunk 13]
```diff
@@ -29,42 +30,35 @@ class GrpcTest {
...
-        assertTrue(testServerOne.storage.blocks.contains(blocks.first().block))
+        assertEquals(GENESIS_SEQ, blocks.first().sequenceNumber)
+        assertTrue(testServerOne.storage.blocks.contains(blocks.first()))
     }
 
-    @Ignore("this should error since it is now a bad weather test and should not deadlock")
-    @Test
-    fun send_half_block_after_crawl() {
+    @Test(expected = Exception::class)
+    fun send_random_crap() {
```
Comment (GijsWeterings) [inline app/src/test/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServerTest.kt:41]: :ok_hand: :see_no_evil:

**Review by GijsWeterings — CHANGES_REQUESTED** [review]
Something is still wrong with this changeset.... I keep getting this exception on startup
EDIT: D'oh, eduroam ;)

```
01-08 13:22:47.349 4041-4041/nl.tudelft.cs4160.identitychain E/AndroidRuntime: FATAL EXCEPTION: main
                                                                               Process: nl.tudelft.cs4160.identitychain, PID: 4041
                                                                               java.lang.RuntimeException: Cannot create an instance of class nl.tudelft.cs4160.identitychain.main.MainViewModel
                                                                                   at android.arch.lifecycle.ViewModelProviders$DefaultFactory.create(ViewModelProviders.java:156)
                                                                                   at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:130)
                                                                                   at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:98)
                                                                                   at nl.tudelft.cs4160.identitychain.main.PeerConnectFragment.onCreateView(PeerConnectFragment.kt:30)
                                                                                   at android.support.v4.app.Fragment.performCreateView(Fragment.java:2354)
                                                                                   at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1419)
                                                                                   at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1740)
                                                                                   at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1809)
                                                                                   at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:799)
                                                                                   at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2580)
                                                                                   at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2367)
                                                                                   at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2322)
                                                                                   at android.support.v4.app.FragmentManagerImpl.execSingleAction(FragmentManager.java:2199)
                                                                                   at android.support.v4.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:651)
                                                                                   at android.support.v4.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:145)
                                                                                   at android.support.v4.view.ViewPager.populate(ViewPager.java:1236)
                                                                                   at android.support.v4.view.ViewPager.populate(ViewPager.java:1084)
                                                                                   at android.support.v4.view.ViewPager.onMeasure(ViewPager.java:1614)
                                                                                   at android.view.View.measure(View.java:22104)
                                                                                   at android.support.constraint.ConstraintLayout.internalMeasureChildren(ConstraintLayout.java:934)
                                                                                   at android.support.constraint.ConstraintLayout.onMeasure(ConstraintLayout.java:973)
                                                                                   at android.view.View.measure(View.java:22104)
                                                                                   at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6584)
                                                                                   at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
                                                                                   at android.support.v7.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:139)
                                                                                   at android.view.View.measure(View.java:22104)
                                                                                   at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6584)
                                                                                   at android.support.v7.widget.ActionBarOverlayLayout.onMeasure(ActionBarOverlayLayout.java:400)
                                                                                   at android.view.View.measure(View.java:22104)
                                                                                   at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6584)
                                                                                   at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
                                                                                   at android.view.View.measure(View.java:22104)
                                                                                   at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6584)
                                                                                   at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1514)
                                                                                   at android.widget.LinearLayout.measureVertical(LinearLayout.java:806)
                                                                                   at android.widget.LinearLayout.onMeasure(LinearLayout.java:685)
                                                                                   at android.view.View.measure(View.java:22104)
                                                                                   at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6584)
                                                                                   at android.widget.FrameLayout.onMeasure(FrameLayout.java:185)
                                                                                   at com.android.internal.policy.DecorView.onMeasure(DecorView.java:722)
                                                                                   at android.view.View.measure(View.java:22104)
                                                                                   at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:2441)
                                                                                   at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:1522)
                                                                                   at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1775)
                                                                                   at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1410)
                                                                                   at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6834)
                                                                                   at android.view.Choreographer$CallbackRecord.run(Choreographer.java:966)
                                                                                   at android.view.Choreographer.doCallbacks(Choreographer.java:778)
                                                                                   at android.view.Choreographer.doFrame(Choreographer.java:713)
                                                                                   at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:952)
                                                                                   at android.os.Handler.handleCallback(Handler.java:789)
01-08 13:22:47.350 4041-4041/nl.tudelft.cs4160.identitychain E/AndroidRuntime:     at android.os.Handler.dispatchMessage(Handler.java:98)
                                                                                   at android.os.Looper.loop(Looper.java:164)
                                                                                   at android.app.ActivityThread.main(ActivityThread.java:6809)
                                                                                   at java.lang.reflect.Method.invoke(Native Method)
                                                                                   at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
                                                                                   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
                                                                                Caused by: java.lang.reflect.InvocationTargetException
                                                                                   at java.lang.reflect.Constructor.newInstance0(Native Method)
                                                                                   at java.lang.reflect.Constructor.newInstance(Constructor.java:334)
                                                                                   at android.arch.lifecycle.ViewModelProviders$DefaultFactory.create(ViewModelProviders.java:148)
                                                                                   	... 56 more
                                                                                Caused by: kotlin.KotlinNullPointerException
                                                                                   at nl.tudelft.cs4160.identitychain.main.MainViewModel.<init>(MainViewModel.kt:60)
                                                                                   	... 59 more
```

File: app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt (hunk 8 above)
Comment (Griezzly) [reply app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:209]: no just the publicKey as a ChainService.Key (proto stuff) that we get back from the message. else that would mean we send the private key aswell.

**Pull request merged**